File: backend/src/main/java/es/udc/paproject/backend/model/services/ExerciseService.java
package es.udc.paproject.backend.model.services;

import java.util.List;

import es.udc.paproject.backend.model.entities.Exercise;
import es.udc.paproject.backend.model.exceptions.InstanceNotFoundException;
import es.udc.paproject.backend.model.exceptions.UsedExerciseException;

public interface ExerciseService {
    
    Exercise addExercise(String exerciseName, String description, String objective, String exerciseType) throws InstanceNotFoundException;

    void addExerciseToTraining(Long trainingId, Long exerciseId) throws InstanceNotFoundException;

    void addExerciseToGame(Long gameId, Long exerciseId) throws InstanceNotFoundException;

    Exercise findExerciseById(Long exerciseId) throws InstanceNotFoundException;

    List<Exercise> findAllExercises() throws InstanceNotFoundException;

    Block<Exercise> findAllExercises(int page, int size) throws InstanceNotFoundException;

    Block<Exercise> findExercisesByType(String exerciseType, int page, int size) throws InstanceNotFoundException;
    
    List<Exercise> findExercisesByType(String exerciseType) throws InstanceNotFoundException;

    List<Exercise> findExercisesByTrainingId(Long trainingId) throws InstanceNotFoundException;

    List<Exercise> findExercisesByGameId(Long gameId) throws InstanceNotFoundException;

    void removeExercise(Long exerciseId) throws InstanceNotFoundException, UsedExerciseException;

    void removeExerciseToTraining(Long exerciseId, Long trainingId) throws InstanceNotFoundException;

    void removeExerciseToGame(Long exerciseId, Long gameId) throws InstanceNotFoundException;

    Exercise updatExercise(Long exerciseId, String exerciseName, String description, String objective, String exerciseType) throws InstanceNotFoundException;
}

File: frontend/src/backend/exerciseService.js
import {config, appFetch} from './appFetch';

export const findExerciseById = (exerciseId, onSuccess, onErrors) => 
    appFetch(`/exercises/${exerciseId}`, config('GET'), onSuccess, onErrors);

export const findAllExercises = (onSuccess, onErrors) => 
    appFetch('/exercises', config('GET'), onSuccess, onErrors);

export const findAllExercisesPage = ({page}, onSuccess, onErrors) => 
    appFetch(`/exercises/page?page=${page}`, config('GET'), onSuccess, onErrors);

export const findExercisesByTypePage = ({exerciseType, page}, onSuccess, onErrors) =>
    appFetch(`/exercises/${exerciseType}/exerciseType/page?page=${page}`, config('GET'), onSuccess, onErrors);

export const findExercisesByType = (exerciseType, onSuccess, onErrors) =>
    appFetch(`/exercises/${exerciseType}/exerciseType`, config('GET'), onSuccess, onErrors);

export const findExercisesByTrainingId = (trainingId, onSuccess, onErrors) => 
    appFetch(`/exercises/${trainingId}/training`, config('GET'), onSuccess, onErrors);

export const findExercisesByGameId = (gameId, onSuccess, onErrors) => 
    appFetch(`/exercises/${gameId}/game`, config('GET'), onSuccess, onErrors);

export const addExercise = (exerciseName, description, objective, exerciseType, onSuccess, onErrors) => 
    appFetch(`/exercises?exerciseName=${exerciseName}&description=${description}&objective=${objective}&exerciseType=${exerciseType}`, config('POST'), onSuccess, onErrors);

export const addExerciseToTraining = (trainingId, exerciseId, onSuccess, onErrors) =>
    appFetch(`/exercises/${trainingId}/addExerciseToTraining?exerciseId=${exerciseId}`, config('POST'), onSuccess, onErrors);

export const addExerciseToGame = (gameId, exerciseId, onSuccess, onErrors) =>
    appFetch(`/exercises/${gameId}/addExerciseToGame?exerciseId=${exerciseId}`, config('POST'), onSuccess, onErrors);

export const updatExercise = (exerciseId, exerciseName, description, objective, exerciseType, onSuccess, onErrors) => 
    appFetch(`/exercises/${exerciseId}?exerciseName=${exerciseName}&description=${description}&objective=${objective}&exerciseType=${exerciseType}`, config('PUT'), onSuccess, onErrors);

export const removeExercise = (exerciseId, onSuccess, onErrors) => 
    appFetch(`/exercises/${exerciseId}`, config('DELETE'), onSuccess, onErrors);

export const removeExerciseToTraining = (trainingId, exerciseId, onSuccess, onErrors) =>
    appFetch(`/exercises/${trainingId}/training?exerciseId=${exerciseId}`, config('DELETE'), onSuccess, onErrors);

export const removeExerciseToGame = (gameId, exerciseId, onSuccess, onErrors) =>
    appFetch(`/exercises/${gameId}/game?exerciseId=${exerciseId}`, config('DELETE'), onSuccess, onErrors);